### PR TITLE
Fix fh in imputer method based on in-sample forecasts

### LINKED ISSUE #859: ARIMA and AutoARIMA should return pandas.NA when value cannot be predicted, e.g., first index in-sample for start=0, d=1
Update 2021-09-15 - re-opened since a recurring issue.

Not a bug, but an interface convention - currently ARIMA, AutoARIMA and other learners fail when asked for a prediction that is undefined, e.g., difference parameter d=1 and first in-sample index (start=0).

It would be more sensible if `pandas.NA` is returned instead for such indices, perhaps a warning should be raised too that `pandas.NA` are produced.

In first instance, we should update ARIMA and AutoARIMA with this convention.

---

**Describe the bug**

When using `Imputer` with `forecaster = AutoARIMA()`  the error "ValueError: In-sample predictions undefined for start=0 when d=1" is raised.

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-2d05f4822fd6> in <module>()
      4 
      5 transformer = Imputer(method="forecaster", forecaster=AutoARIMA()).fit(Z=df.value)
----> 6 y_hat = transformer.transform(Z=df.value)

/usr/local/lib/python3.7/dist-packages/sktime/transformations/series/impute.py in transform(self, Z, X)
    102             z_pred = forecaster.fit(
    103                 z.fillna(method="ffill").fillna(method="backfill")
--> 104             ).predict(fh=fh_ins)
    105             # fill with trend values
    106             z = z.fillna(value=z_pred)

/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/_sktime.py in predict(self, fh, X, return_pred_int, alpha)
    226         self.check_is_fitted()
    227         self._set_fh(fh)
--> 228         return self._predict(self.fh, X, return_pred_int=return_pred_int, alpha=alpha)
    229 
    230     def compute_pred_int(self, y_pred, alpha=DEFAULT_ALPHA):

/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/adapters/_pmdarima.py in _predict(self, fh, X, return_pred_int, alpha)
     58         # all values are in-sample
     59         elif fh.is_all_in_sample(self.cutoff):
---> 60             return self._predict_in_sample(fh_ins, **kwargs)
     61 
     62         # both in-sample and out-of-sample values

/usr/local/lib/python3.7/dist-packages/sktime/forecasting/base/adapters/_pmdarima.py in _predict_in_sample(self, fh, X, return_pred_int, alpha)
     80             X=X,
     81             return_conf_int=return_pred_int,
---> 82             alpha=alpha,
     83         )
     84 

/usr/local/lib/python3.7/dist-packages/pmdarima/utils/metaestimators.py in <lambda>(*args, **kwargs)
     51 
     52         # lambda, but not partial, allows help() to work with update_wrapper
---> 53         out = (lambda *args, **kwargs: self.fn(obj, *args, **kwargs))
     54         # update the docstring of the returned function
     55         update_wrapper(out, self.fn)

/usr/local/lib/python3.7/dist-packages/pmdarima/arima/auto.py in predict_in_sample(self, X, start, end, dynamic, return_conf_int, alpha, typ, **kwargs)
    242             return_conf_int=return_conf_int,
    243             alpha=alpha,
--> 244             typ=typ,
    245         )
    246 

/usr/local/lib/python3.7/dist-packages/pmdarima/arima/arima.py in predict_in_sample(self, X, start, end, dynamic, return_conf_int, alpha, **kwargs)
    584         if start is not None and start < d:
    585             raise ValueError("In-sample predictions undefined for start={0} "
--> 586                              "when d={1}".format(start, d))
    587 
    588         # if we fit with exog, make sure one was passed:

ValueError: In-sample predictions undefined for start=0 when d=1
```

**To Reproduce**

Data:
[df.csv](https://github.com/alan-turing-institute/sktime/files/6426660/df.csv) 

```python
import pandas as pd
from sktime.transformations.series.impute import Imputer
from sktime.forecasting.arima import AutoARIMA

df = pd.read_csv('df.csv')  
df['timestamp'] = pd.DatetimeIndex(df['timestamp']).to_period('D')
df = df.set_index('timestamp')

transformer = Imputer(method="forecaster", forecaster=AutoARIMA()).fit(Z=df.value)
y_hat = transformer.transform(Z=df.value)
```

**Expected behavior**

`transformer.transform(Z)` should return `Z` with NAn values replaced by AutoArima predictions.

**Additional context**

I think the problem is in `/usr/local/lib/python3.7/dist-packages/sktime/transformations/series/impute.py` in `transform(self, Z, X)`, line 100, because all the points are being passed to `fh_ins` instead of just the points with NaN.

```python
100             fh_ins = -np.arange(len(z))
101             # fill NaN before fitting with ffill and backfill (heuristic)
102             z_pred = forecaster.fit(
103                 z.fillna(method="ffill").fillna(method="backfill")
104             ).predict(fh=fh_ins)
105             # fill with trend values
106             z = z.fillna(value=z_pred)

```

**Versions**
<details>

```
System:
    python: 3.7.10 (default, Feb 20 2021, 21:17:23)  [GCC 7.5.0]
executable: /usr/bin/python3
   machine: Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic

Python dependencies:
          pip: 19.3.1
   setuptools: 56.0.0
      sklearn: 0.24.2
       sktime: 0.6.0
  statsmodels: 0.12.2
        numpy: 1.19.5
        scipy: 1.4.1
       Cython: 0.29.22
       pandas: 1.1.5
   matplotlib: 3.2.2
       joblib: 1.0.1
        numba: 0.51.2
     pmdarima: 1.8.2
      tsfresh: None
```

</details>

### PR DESCRIPTION
**Pull request by julramos**
#### Reference Issues/PRs

Fixes #859.

#### What does this implement/fix? Explain your changes.

When using a forecaster to impute data, the class `Imputer` was passing all the points to `forecaster.predict()`. This could throw the error "ValueError: In-sample predictions undefined for start=0 when d=1" when using `forecaster = AutoARIMA()`. I changed the code to pass only the points containing NaN to `forecaster.predict()`.

#### Does your contribution introduce a new dependency? If yes, which one?

No.

#### What should a reviewer concentrate their feedback on?

Nothing in particular.

#### Any other comments?

No. 

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

### CODE REVIEW

File: sktime/transformations/series/impute.py
Comment (mloning) [inline sktime/transformations/series/impute.py:121]: Can we not simplify this when using the `ForecastingHorizon` class from `sktime.forecasting.base`?

**Review by mloning — COMMENTED** [review]
Thanks @julramos for the PR - much appreciated! Make sure you add yourself to the list of contributors.

File: sktime/transformations/series/impute.py
Comment (julramos) [reply sktime/transformations/series/impute.py:121]: I didn't understange what you suggested exactly. Is it something like this?

```python
            # fill NaN before fitting with ffill and backfill (heuristic)
            Z_aux = Z.fillna(method="ffill").fillna(method="backfill")
            # multivariate
            if isinstance(Z_aux, pd.DataFrame):
                for col in Z_aux:
                    forecaster.fit(y=Z_aux[col])
                    # in-sample forecasting horizon
                    fh_ins = ForecastingHorizon(values=np.where(Z[col].isna())[0], is_relative=False)
                    fh_ins = fh_ins.to_relative(cutoff=len(Z)-1).to_numpy()
                    Z_pred = forecaster.predict(fh=fh_ins)
                    Z[col] = Z[col].fillna(value=Z_pred)
            # univariate
            else:
                forecaster.fit(y=Z_aux)
                # in-sample forecasting horizon
                fh_ins = ForecastingHorizon(values=np.where(Z.isna())[0], is_relative=False)
                fh_ins = fh_ins.to_relative(cutoff=len(Z)-1).to_numpy()
                Z_pred = forecaster.predict(fh=fh_ins)
                Z = Z.fillna(value=Z_pred)
```

Comment (mloning) [reply sktime/transformations/series/impute.py:121]: Sorry for the slow reply @julramos, I think you can simplify it even further, here's a simple example:
```python
y = load_airline()
y.iloc[np.random.randint(0, len(y), 20)] = np.nan

forecaster = AutoARIMA(d=0, suppress_warnings=True)

isna = y.isna()
fh = ForecastingHorizon(y.index[isna], is_relative=False)  # define fh based on index of missing values 
y_aux = y.fillna(method="ffill").fillna(method="backfill")
forecaster.fit(y_aux, fh=fh) 
y_pred = forecaster.predict()
y.iloc[isna] = y_pred  # replace missing values with predicted values
```

A few remarks:
* make sure to pass `fh` to fit rather than predict (some forecaster need the horizon already during fitting)
* in the above if-else statement there seems to be some code duplication, could we write a function for univariate imputation based on the forecaster and then loop over it with the column selection?

File: sktime/transformations/series/impute.py
Comment (mloning) [inline sktime/transformations/series/impute.py:170]: This looks much better already. I would further try to improve this. Here are some suggestions:
* make `_univariate_forecaster` a stand-alone function rather than a method (functions are simpler than methods)
* rename to `_impute_with_forecaster(...)` (more descriptive)
* change signature to `_impute_with_forecaster(forecaster, z)`
* keep if statements conditional on `self.method` in `transform` method rather than moving it inside the function
* move loop over columns if we have multivariate data (`pd.Dataframe`) into the `_impute_with_forecaster(...)` function (applies to all forecasters)

So something along these lines:
```python
def transform(...):
    ...
    if self.method == "drift":
        forecaster = PolynomialTrendForecaster(...)
        Z = _impute_with_forecaster(forecaster, Z)
    if self.method == "forecaster":
        forecaster = clone(self.forecaster)
        Z = _impute_with_forecaster(forecaster, Z)
    ...
```
Would that work?

**Comment by mloning** [comment]
Hi @julramos - we introduced more stringent docstring tests yesterday (#752), could you try to fix the docstrings? You can `pip install pydocstyle` and run `pydocstyle <file>` to check if the tests pass locally. Let me know if you have any questions.

**Review by mloning — DISMISSED** [review]
Thanks again @julramos, I'm happy to merge this now. @aiwalter any last comments before we merge?

**Review by aiwalter — COMMENTED** [review]
Just left few comments @julramos

File: sktime/transformations/series/impute.py
Comment (mloning) [reply sktime/transformations/series/impute.py:182]: It's a private function, so it seems we don't enforce docstring formatting for those.

File: sktime/transformations/series/impute.py
Comment (mloning) [reply sktime/transformations/series/impute.py:3]: What do you mean? We need a docstring at the top of public files I think.

**Comment by mloning** [comment]
@aiwalter is it okay if we merge this? I think your comments are minor changes, if you think they are important, we can also implement them quickly before merging.

**Review by mloning — APPROVED** [review]
Hi @julramos and @aiwalter, I implemented the suggested changes and will merge now once all tests pass unless you have any final comments!

**Comment by mloning** [comment]
Now merged - thanks again @julramos 🎉

**Pull request merged**
